pkg/provision/workspace/persistentHome.go

```diff
@@ -0,0 +1,87 @@
+//
+// Copyright (c) 2019-2023 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package workspace
+
+import (
+	"fmt"
+
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+
+	"github.com/devfile/devworkspace-operator/pkg/common"
+	"github.com/devfile/devworkspace-operator/pkg/constants"
+	"github.com/devfile/devworkspace-operator/pkg/dwerrors"
+)
+
+// Adds a Devfile volume to the given DevWorkspace and mounts the volume to `/home/user/` for every
+// container component defined in the DevWorkspace by adding additional volumeMounts.
+// Has no effect if `persistUserHome` is disabled in the DevWorkspaceOperatorConfig.
+// Returns an error if `persistUserHome` is enabled in the DWOC and at least one of the container components
+// in the DevWorkspace already mounts to `/home/user/`.
+func ProvisionPersistentUserHome(workspace *common.DevWorkspaceWithConfig) error {
+	shouldPersistHome, err := needsPersistentHomeDirectory(workspace)
+	if err != nil {
+		return err
+	}
+	if !shouldPersistHome {
+		return nil
+	}
+
+	homeVolume := v1alpha2.Component{
+		Name: constants.HomeVolumeName,
+		ComponentUnion: v1alpha2.ComponentUnion{
+			Volume: &v1alpha2.VolumeComponent{},
+		},
+	}
+	homeVolumeMount := v1alpha2.VolumeMount{
+		Name: constants.HomeVolumeName,
+		Path: constants.HomeUserDirectory,
+	}
+
+	workspace.Spec.Template.Components = append(workspace.Spec.Template.Components, homeVolume)
+	for _, component := range workspace.Spec.Template.Components {
+		if component.Container == nil {
+			continue
+		}
+		component.Container.VolumeMounts = append(component.Container.VolumeMounts, homeVolumeMount)
+	}
```

Comment:
Does it make sense to do an extra-careful look before we leap here? E.g. apply the changes to a copy of `components`, check that the resulting workspace is still valid via `devfilevalidation.ValidateComponents(components)` and only return the added home if it doesn't break the workspace?

My main concern here is applying a change that takes a valid devfile and makes it invalid through some edge case (e.g. if the user has a volume named `persistentHome` that mounts somewhere else).

Reply:
+1 good call, I added the use of `devfilevalidation.ValidateComponents(components)` and added an extra test for the case where the user already has a volume named `persistentHome`